### DISCUSSION TITLE
`ConfigurationBinder.Bind` and `ConfigurationBinder.Get` trimming warning

### DISCUSSION THREAD

**Discussion by eduherminio** [dotnet/runtime#59230] [Q&A] [answered]
Given a project with :

```xml
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <PublishTrimmed>true</PublishTrimmed>
    <TrimMode>link</TrimMode>
  </PropertyGroup>
```

```csharp
public class GeneralSettings {}

var settingsInstance = new GeneralSettings();
var config = new ConfigurationBuilder().Build();
```

When I use `ConfigurationBinder.Bind`

```csharp
config.GetRequiredSection(nameof(EngineSettings)).Bind(settingsInstance);
```

I get the following warning:

```bash
warning IL2026:
Using method 'Microsoft.Extensions.Configuration.ConfigurationBinder.Bind(Microsoft.Extensions.Configuration.IConfiguration, object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Cannot statically analyze the type of instance so its members may be trimmed.
```

If I use `ConfigurationBuilder.Get` I also get a the same one:

```csharp
settingsInstance = config.GetRequiredSection(nameof(GeneralSettings)).Get<GeneralSettings>();
```

```bash
Using method 'Microsoft.Extensions.Configuration.ConfigurationBinder.Get<Lynx.GameParameters>(Microsoft.Extensions.Configuration.IConfiguration)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. In case the type is non-primitive, the trimmer cannot statically analyze the object's type so its members may be trimmed.
```

In both cases the code works after the trimming.

Should I suppress the warning?
Should I 'customize' the trimming?

I've read  [Customizing Trimming in .NET 5](https://devblogs.microsoft.com/dotnet/customizing-trimming-in-net-core-5), but I'm unsure what's the best way to proceed for these specific methods.

**Comment by davidfowl** [comment, marked as answer]
These methods aren't trimming friendly. The linker can't tell what properties it should keep when you use these methods. Your options are:
- Bind the value manually (don't use these methods)
- The containing methods can declare what properties of `GeneralSettings` are used (using the dynamic dependency annotation)
- You can use the xml file to customize the linking process to preserve `GeneralSettings` properties.

**Reply by eduherminio** [reply]
Could you elaborate on the second option? I'm not sure to follow that one,

**Reply by davidfowl** [reply]
You're fine and you can suppress the warning. The linker can't tell what if the type being passed into Bind has been trimmed so it complains for everything. In this case it works because application code isn't trimmed.

**Reply by molinch** [reply]
@davidfowl Do you know if the linker will be improved to not pop in that case: when app code isn't trimmed but classes are passed to methods which have RequiresUnreferencedCodeAttribute?

**Reply by davidfowl** [reply]
Unfortunately this overload isn't something the linker understands. Usually overloads like this take a Type or have a generic type argument that the linker then does flow analysis on. This API unfortunately doesn't take the type information so the linker doesn't know what to do with it.

**Reply by molinch** [reply]
Thanks @davidfowl, while for `Bind(object)` I get it since the signature expects an object. However when using `.Get<T>()` the type is known so the linker should not pop warnings ideally. Do you agree?

**Reply by davidfowl** [reply]
`Get<T>` is currently marked this way for the same reason that `JsonSerializer.Serialize<T>` is marked as `RequiresUnreferencedCodeAttribute`. The logic in these calls can't easily be described using the attributes we have to educate the linker. It really comes down to abstractions. When you have something like this:

```
public class GeneralSettings 
{
    public IEnumerable<string> Values { get; set; }
}
```

What types should the linker preserve for values? string[]? `List<string>`? etc etc. The "fix" for these situations is to design a source generator that can describe these hints or generate the code to do the serialization itself (that's the approach we're exploring).

**Reply by molinch** [reply]
Great explanation, thanks @davidfowl 👍

**Reply by jetersen** [reply]
Anyone who managed to manually bind? I wonder how to best achieve this 😓

**Reply by jasonterando** [reply]
Something like this is working for me...

```c#
  services.AddSingleton<MyConfigurationModel>(provider =>
  {
      var c = provider
          .GetRequiredService<IConfiguration>()
          .GetRequiredSection("MyConfiguration");
      var config = new MyConfigurationModel
          c["Url"] ?? "https://default",
          c["UserName"] ?? ""
      );
      config.Validate();  // member method to validate parameters
      return config;
  });
```

**Reply by davidfowl** [reply]
We're working on a source generator for configuration binding. Until then, manually binding really means manually binding each property by hand.

**Reply by jasonterando** [reply]
@davidfowl  - great to hear!  Source generators FTW

**Reply by johnwc** [reply]
> We're working on a source generator for configuration binding. Until then, manually binding really means manually binding each property by hand.

Or we can safely suppress the warning for now?

**Reply by colejohnson66** [reply]
@johnwc It is generally not safe to silence trim warnings.

**Reply by johnwc** [reply]
@colejohnson66 yes, I am aware of that and follow that rule. But, @davidfowl stated previously to silence it and I wanted to validate from him that it is ok to do such a thing.

**Reply by ThaDaVos** [reply]
Any update on the source generators?

**Reply by jasper-d** [reply]
> Any update on the source generators?

It is part of .NET 8 but you'll need to enable it explicitly.

https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8/runtime#configuration-binding-source-generator
